src/System.Private.CoreLib/shared/System/SpanHelpers.Char.cs

```diff
@@ -10,10 +10,56 @@
 using Internal.Runtime.CompilerServices;
 #endif
 
+#if BIT64
+using nuint = System.UInt64;
+#else
+using nuint = System.UInt32;
+#endif
+
 namespace System
 {
     internal static partial class SpanHelpers // .Char
     {
+        public static int IndexOf(ref char searchSpace, int searchSpaceLength, ref char value, int valueLength)
+        {
+            Debug.Assert(searchSpaceLength >= 0);
+            Debug.Assert(valueLength >= 0);
+
+            if (valueLength == 0)
+                return 0;  // A zero-length sequence is always treated as "found" at the start of the search space.
+
+            char valueHead = value;
+            ref char valueTail = ref Unsafe.Add(ref value, 1);
+            int valueTailLength = valueLength - 1;
+
+            int index = 0;
+            for (; ; )
+            {
+                Debug.Assert(0 <= index && index <= searchSpaceLength); // Ensures no deceptive underflows in the computation of "remainingSearchSpaceLength".
+                int remainingSearchSpaceLength = searchSpaceLength - index - valueTailLength;
+                if (remainingSearchSpaceLength <= 0)
+                    break;  // The unsearched portion is now shorter than the sequence we're looking for. So it can't be there.
+
+                // Do a quick search for the first element of "value".
+                int relativeIndex = IndexOf(ref Unsafe.Add(ref searchSpace, index), valueHead, remainingSearchSpaceLength);
+                if (relativeIndex == -1)
+                    break;
+                index += relativeIndex;
+
+                // Found the first element of "value". See if the tail matches.
+                if (SequenceEqual(
+                    ref Unsafe.As<char, byte>(ref Unsafe.Add(ref searchSpace, index + 1)),
```

Comment:
Is it ever possible for _index_ to be `Int32.MaxValue`, causing this addition to integer overflow? I _think_ the answer is no (because `IndexOf` should never return that value) but am having trouble convincing myself of this.

Reply:
Switched it up for you https://github.com/dotnet/coreclr/pull/21076/commits/2252666fc40f1ba85ed6cf87b76fc2a4ab863595

Reply:
Also the `byte` version